backend/api/login.go

```diff
@@ -1,99 +1,90 @@
-	if ok || user_id != nil {
-		http.Redirect(w, r, UrlFor("timeline", ""), http.StatusFound)
-		return
-	}
+	username := body.Username
+	pwd := body.PWD
 
+	user, ok := s.userRepo.GetUser(username)
 	var errorStr *string = nil
 
-	if r.Method == "POST" {
-		err := r.ParseForm()
+	if !ok || user == nil {
+		s := "Invalid username"
+		errorStr = &s
+	} else if !util.CheckPassword(pwd, user.Pw_hash) {
+		s := "Invalid password"
+		errorStr = &s
+	} else {
+		session, err := s.GetStore().Get(r, "auth")
+
 		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-		}
-		vals := r.PostForm
-		user, ok := s.userRepo.GetUser(vals.Get("username"))
-
-		if !ok || user == nil {
-			s := "Invalid username"
-			errorStr = &s
-		} else if !util.CheckPassword(vals.Get("password"), user.Pw_hash) {
-			s := "Invalid password"
-			errorStr = &s
-		} else {
-			session, err := s.GetStore().Get(r, "auth")
-
-			if err != nil {
-				log.Println("Error getting session:", err)
-				w.WriteHeader(http.StatusInternalServerError)
-				return
-			}
-
-			session.Values["user"] = user.User_id
-			err = session.Save(r, w)
-
-			if err != nil {
-				log.Println("Error logging in:", err)
-				w.WriteHeader(http.StatusInternalServerError)
-				return
-			}
-
-			s.PushFlashMessage(w, r, "You were logged in")
-			http.Redirect(w, r, UrlFor("timeline", ""), http.StatusFound)
+			log.Println("Error getting session:", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
 		}
-	}
 
-	t, err := template.New("layout.html").Funcs(s.GetFuncMap()).ParseFiles("../web/templates/layout.html", "../web/templates/login.html")
+		session.Values["user"] = user.User_id
+		err = session.Save(r, w)
 
-	if err != nil {
-		log.Println("Error creating template:", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
+		if err != nil {
+			log.Println("Error logging in:", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 	}
 
-	data := model.Template{
-		Request: model.RenderRequest{Endpoint: "login"},
-		Error:   errorStr,
-		Flashes: s.GetFlashedMessages(w, r),
-	}
+	if errorStr != nil {
+		t := ErrReturn{Status: http.StatusBadRequest, ErrorMsg: *errorStr}
+		m, err := json.Marshal(t)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 
-	if err = t.Execute(w, data); err != nil {
-		log.Println("Error rendering frontend:", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write(m)
+	} else {
+		w.WriteHeader(http.StatusOK)
 	}
 }
 
 func (s *Server) LogoutHandler(w http.ResponseWriter, r *http.Request) {
 	session, err := s.GetStore().Get(r, "auth")
+
 	if err != nil {
 		log.Println("Error getting session", err)
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
+
 	for k := range session.Values {
 		delete(session.Values, k)
 	}
+	session.Options.MaxAge = -1
```

Comment:
What does this do?

Reply:
It removes the cookie by immediately expiring it, as the `delete` for some reason doesn't. Maybe you got a clue why

Reply:
My idea was that the delete just deletes the info in the `Auth` cookie, but it still leaves empty state there. Frontend can't really distinguish between if the Auth cookie has user id or not inside